.circleci/config.yml

```diff
@@ -31,15 +29,51 @@ jobs:
       - run:
           name: Start LDAP
           command: sudo /etc/init.d/slapd stop
-      - run:
-          name: save SHA to a file
-          command: echo $CIRCLE_SHA1 > .circle-sha
       - restore_cache:
           keys:
-            - v1-repo-{{ checksum ".circle-sha" }}
+            - v1-repo-{{ .Environment.CIRCLE_SHA1 }}
       - run: go get -u github.com/golang/lint/golint
       - run: make testall
 
+  build_binary:
+    docker:
+      - image: circleci/golang:1.8
+    working_directory: /go/src/github.com/coreos/dex
+    steps:
+      - restore_cache:
+          keys:
+            - v1-repo-{{ .Environment.CIRCLE_SHA1 }}
+      - run:
+          name: Build dex binary
+          command: |
+            make release-binary
+            mkdir -p _output/bin/dex && cp /go/bin/dex _output/bin/dex
+      - save_cache:
+          key: v1-binary-{{ .Environment.CIRCLE_SHA1 }}
+          paths:
+            - /go/src/github.com/coreos/dex/_output/bin
+
+  deploy_image:
+    docker:
+      - image: circleci/golang:1.8
+    working_directory: /go/src/github.com/coreos/dex
+    steps:
+      - restore_cache:
+          keys:
+            - v1-repo-{{ .Environment.CIRCLE_SHA1 }}
+      - restore_cache:
+          keys:
+            - v1-binary-{{ .Environment.CIRCLE_SHA1 }}
+      - setup_remote_docker
+      - deploy:
+          name: Install Docker client
+          command: |
+            set -x
+            sudo curl -fsSLO https://get.docker.com/builds/Linux/x86_64/docker-17.05.0-ce.tgz && sudo tar --strip-components=1 -xvzf docker-17.05.0-ce.tgz -C /usr/local/bin
+      - deploy:
+          name: Deploy dex Docker image
+          command: /bin/bash -e scripts/circle-deploy
```

Comment:
do we still need `/bin/bash -e`? since `scripts/circle-deploy` already has the shebang on top that declares that